File: sdks/code-interpreter/python/src/code_interpreter/sync/code_interpreter.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
Synchronous Code Interpreter SDK.
"""

import logging
import time
from datetime import timedelta

from opensandbox.constants import DEFAULT_EXECD_PORT
from opensandbox.exceptions import (
    InvalidArgumentException,
    SandboxException,
    SandboxInternalException,
    SandboxReadyTimeoutException,
)
from opensandbox.sync.sandbox import SandboxSync

from code_interpreter.sync.adapters.factory import AdapterFactorySync
from code_interpreter.sync.services.code import CodesSync

logger = logging.getLogger(__name__)

DEFAULT_READY_TIMEOUT = timedelta(seconds=30)
DEFAULT_HEALTH_CHECK_POLLING_INTERVAL = timedelta(milliseconds=200)

# Strict health check script: verifies the code interpreter runtime (Jupyter
# kernel gateway) is actually serving inside the sandbox. execd starts serving
# /ping before the entrypoint launches Jupyter, and the setup stage may run
# short-lived "jupyter kernelspec" helpers, so a daemon ping or a process-name
# grep cannot prove the runtime is ready. Probing the Jupyter listen port
# (127.0.0.1:${JUPYTER_PORT:-44771}, same default as the entrypoint) only
# passes once the server accepts connections.
RUNTIME_PROCESS_CHECK_COMMAND = (
    "bash -c 'exec 3<>/dev/tcp/127.0.0.1/${JUPYTER_PORT:-44771}' "
    "&& exit 0 || exit 1"
)


class CodeInterpreterSync:
    """
    Synchronous Code Interpreter SDK providing secure, isolated code execution capabilities.

    This class mirrors the async :class:`code_interpreter.code_interpreter.CodeInterpreter`, but all
    operations are **blocking** and executed in the current thread.

    It wraps an existing :class:`opensandbox.sync.sandbox.SandboxSync` instance and adds
    code-execution APIs (contexts, run with SSE streaming, interrupts) on top.

    Notes:

    - **Blocking**: Do not call these methods directly from an asyncio event loop thread.
      If you need non-blocking behavior, prefer the async :class:`~code_interpreter.code_interpreter.CodeInterpreter`.
    - **Lifecycle**: Remote lifecycle is owned by the underlying sandbox; call methods on
      ``interpreter.sandbox`` for pause/resume/kill/renew/metrics/info/endpoints.

    Usage Example:

    ```python
    from opensandbox.sync.sandbox import SandboxSync
    from code_interpreter.sync.code_interpreter import CodeInterpreterSync
    from code_interpreter.models.code import SupportedLanguage

    sandbox = SandboxSync.create("python:3.11")
    interpreter = CodeInterpreterSync.create(sandbox=sandbox)

    ctx = interpreter.codes.create_context(SupportedLanguage.PYTHON)
    result = interpreter.codes.run("print('hi')", context=ctx)

    sandbox.kill()
    sandbox.close()
    ```
    """

    def __init__(self, sandbox: SandboxSync, code_service: CodesSync) -> None:
        """
        Initialize CodeInterpreterSync with sandbox and code service.

        Note: This constructor is for internal use. Use :meth:`create` instead.

        Args:
            sandbox: Underlying sandbox instance
            code_service: Code execution service implementation (sync)
        """
        self._sandbox = sandbox
        self._code_service = code_service

    @property
    def sandbox(self) -> SandboxSync:
        """
        Provides access to the underlying sandbox instance.

        Returns:
            The underlying sandbox instance
        """
        return self._sandbox

    @property
    def id(self) -> str:
        """
        Gets the unique identifier of this code interpreter (same as underlying sandbox ID).

        Returns:
            ID of the code interpreter/sandbox
        """
        return self._sandbox.id

    @property
    def files(self):
        """
        Provides access to file system operations within the sandbox.

        Returns:
            Service for filesystem manipulation
        """
        return self._sandbox.files

    @property
    def commands(self):
        """
        Provides access to command execution operations.

        Returns:
            Service for command execution
        """
        return self._sandbox.commands

    @property
    def metrics(self):
        """
        Provides access to sandbox metrics and monitoring.

        Returns:
            Service for metrics retrieval
        """
        return self._sandbox.metrics

    @property
    def codes(self) -> CodesSync:
        """
        Provides access to code execution operations (sync).

        This service enables:
        - Multi-language code execution (Python, JavaScript, Bash, etc.)
        - Execution context management with persistent variables
        - Real-time output streaming and interruption capabilities

        Returns:
            Service for advanced code execution with session support
        """
        return self._code_service

    def ping(self) -> bool:
        """
        Check if the code execution service (execd) is responsive.

        Returns:
            True if the code execution service is responsive, False otherwise
        """
        return self._code_service.ping()

    def is_healthy(self) -> bool:
        """
        Check if the code interpreter is healthy (strict check).

        Healthy means both:

        - the code execution service (execd) answers ``GET /ping``; and
        - the code interpreter runtime (Jupyter kernel gateway) is serving
          inside the sandbox, verified by probing its listen port through the
          execd command API.

        Exceptions raised by either leg are treated as unhealthy.

        Returns:
            True if healthy, False otherwise
        """
        try:
            return self.ping() and self._is_runtime_process_alive()
        except Exception:
            return False

    def _is_runtime_process_alive(self) -> bool:
        """
        Check if the code interpreter runtime (Jupyter) is serving.
        """
        try:
            execution = self._sandbox.commands.run(RUNTIME_PROCESS_CHECK_COMMAND)
            return execution.error is None
        except Exception:
            return False

    def check_ready(
        self,
        timeout: timedelta,
        polling_interval: timedelta,
    ) -> None:
        """
        Wait for the code interpreter to pass the strict health check with polling (blocking).

        Raises:
            SandboxReadyTimeoutException: if the health check doesn't pass within timeout
        """
        logger.info(
            f"Waiting for code interpreter {self.id} to pass health check "
            f"(timeout: {timeout.total_seconds()}s)"
        )

        deadline = time.monotonic() + timeout.total_seconds()
        attempt = 0

        while time.monotonic() < deadline:
            attempt += 1
            if self.is_healthy():
                logger.info(
                    f"Code interpreter {self.id} passed health check "
                    f"after {attempt} attempts"
                )
                return

            remaining = deadline - time.monotonic()
            if remaining <= 0:
                break
            time.sleep(min(polling_interval.total_seconds(), remaining))

        raise SandboxReadyTimeoutException(
            f"Code interpreter {self.id} health check timed out after "
            f"{timeout.total_seconds()}s ({attempt} attempts). The code execution "
            f"service (execd) or the interpreter runtime (Jupyter) did not become "
            f"ready. Pass skip_health_check=True to skip this check."
        )

    @classmethod
    def create(
        cls,
        sandbox: SandboxSync,
        *,
        ready_timeout: timedelta = DEFAULT_READY_TIMEOUT,
        health_check_polling_interval: timedelta = DEFAULT_HEALTH_CHECK_POLLING_INTERVAL,
        skip_health_check: bool = False,
    ) -> "CodeInterpreterSync":
        """
        Create a CodeInterpreterSync from an existing SandboxSync instance (blocking).

        By default a strict health check runs before the interpreter is returned:
        the code execution service (execd) must answer ``GET /ping`` AND the
        code interpreter runtime process (Jupyter kernel gateway) must be
        running inside the sandbox, both within ``ready_timeout``. Set
        ``skip_health_check=True`` to opt out.

        Args:
            sandbox: Existing sandbox instance to wrap with code execution capabilities
            ready_timeout: Maximum time to wait for the code execution service health check
            health_check_polling_interval: Time between health check attempts
            skip_health_check: If True, do not wait for the code execution service
                to become ready; the returned interpreter may fail on first use

        Returns:
            CodeInterpreterSync instance wrapping the sandbox

        Raises:
            InvalidArgumentException: If sandbox is not provided
            SandboxException: If creation fails
            SandboxReadyTimeoutException: If the code execution service health check
                times out and ``skip_health_check`` is False
            SandboxInternalException: If internal service initialization fails
        """
        if sandbox is None:
            raise InvalidArgumentException("Sandbox instance must be provided")

        logger.info(f"Creating code interpreter from sandbox: {sandbox.id}")
        factory = AdapterFactorySync(sandbox.connection_config)
        try:
            endpoint = sandbox.get_endpoint(DEFAULT_EXECD_PORT)
            code_service = factory.create_code_execution_service(endpoint)

            interpreter = cls(sandbox, code_service)

            if not skip_health_check:
                interpreter.check_ready(ready_timeout, health_check_polling_interval)

            logger.info(f"Code interpreter {sandbox.id} created successfully")
            return interpreter
        except Exception as e:
            if isinstance(e, SandboxException):
                raise
            raise SandboxInternalException(
                f"Failed to create code interpreter: {e}", cause=e
            ) from e

File: sdks/code-interpreter/csharp/src/OpenSandbox.CodeInterpreter/CodeInterpreter.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Diagnostics;
using OpenSandbox.CodeInterpreter.Factory;
using OpenSandbox.CodeInterpreter.Services;
using OpenSandbox.Config;
using OpenSandbox.Core;
using OpenSandbox.Internal;
using OpenSandbox.Services;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace OpenSandbox.CodeInterpreter;

/// <summary>
/// Strict health check helpers for code interpreters.
/// </summary>
public static class CodeInterpreterHealthCheck
{
    /// <summary>
    /// The shell command run inside the sandbox to verify the interpreter runtime
    /// (Jupyter kernel gateway) is actually serving. execd starts serving
    /// <c>/ping</c> before the entrypoint launches Jupyter, and the setup stage may
    /// run short-lived "jupyter kernelspec" helpers, so a daemon ping or a
    /// process-name grep cannot prove the runtime is ready. Probing the Jupyter
    /// listen port (127.0.0.1:${JUPYTER_PORT:-44771}, same default as the entrypoint)
    /// only passes once the server accepts connections.
    /// </summary>
    public const string RuntimeCheckCommand =
        "bash -c 'exec 3<>/dev/tcp/127.0.0.1/${JUPYTER_PORT:-44771}' && exit 0 || exit 1";
}

/// <summary>
/// Options for creating a code interpreter.
/// </summary>
public class CodeInterpreterCreateOptions
{
    /// <summary>
    /// Gets or sets the adapter factory. If not provided, a default factory is used.
    /// </summary>
    public ICodeInterpreterAdapterFactory? AdapterFactory { get; set; }

    /// <summary>
    /// Gets or sets diagnostics options such as logging.
    /// </summary>
    public SdkDiagnosticsOptions? Diagnostics { get; set; }

    /// <summary>
    /// Gets or sets whether to skip the strict code-executor readiness check.
    /// The returned interpreter may fail on first use if the execd daemon is
    /// not serving yet.
    /// </summary>
    public bool SkipHealthCheck { get; set; }

    /// <summary>
    /// Gets or sets the timeout for the code execution service health check, in seconds.
    /// </summary>
    public int? ReadyTimeoutSeconds { get; set; }

    /// <summary>
    /// Gets or sets the health check polling interval, in milliseconds.
    /// </summary>
    public int? HealthCheckPollingInterval { get; set; }
}

/// <summary>
/// Options for waiting until a code interpreter is ready.
/// </summary>
public class CodeInterpreterWaitUntilReadyOptions
{
    /// <summary>
    /// Gets or sets the timeout in seconds.
    /// </summary>
    public int ReadyTimeoutSeconds { get; set; }

    /// <summary>
    /// Gets or sets the polling interval in milliseconds.
    /// </summary>
    public int PollingIntervalMillis { get; set; }
}

/// <summary>
/// Code interpreter facade for executing code in multiple languages.
/// </summary>
/// <remarks>
/// This class wraps an existing <see cref="Sandbox"/> and provides a high-level API for code execution.
/// Use <see cref="Codes"/> to create contexts and run code.
/// <see cref="Files"/>, <see cref="Commands"/>, and <see cref="Metrics"/> are exposed for convenience
/// and are the same instances as on the underlying <see cref="Sandbox"/>.
/// This type does not own the remote sandbox lifecycle. Call <see cref="Sandbox.KillAsync"/> when you want to terminate
/// the remote instance. Dispose the wrapped <see cref="Sandbox"/> to release local SDK resources.
/// </remarks>
public sealed class CodeInterpreter
{
    /// <summary>
    /// Gets the underlying sandbox instance.
    /// </summary>
    public Sandbox Sandbox { get; }

    /// <summary>
    /// Gets the codes service for code execution operations.
    /// </summary>
    public ICodes Codes { get; }

    /// <summary>
    /// Gets the sandbox ID.
    /// </summary>
    public string Id => Sandbox.Id;

    /// <summary>
    /// Gets the filesystem service.
    /// </summary>
    public ISandboxFiles Files => Sandbox.Files;

    /// <summary>
    /// Gets the command execution service.
    /// </summary>
    public IExecdCommands Commands => Sandbox.Commands;

    /// <summary>
    /// Gets the metrics service.
    /// </summary>
    public IExecdMetrics Metrics => Sandbox.Metrics;

    private readonly ILogger _logger;
    private readonly HttpClientWrapper? _fallbackExecdClient;

    private CodeInterpreter(Sandbox sandbox, ICodes codes, ILogger logger, HttpClientWrapper? fallbackExecdClient = null)
    {
        Sandbox = sandbox ?? throw new ArgumentNullException(nameof(sandbox));
        Codes = codes ?? throw new ArgumentNullException(nameof(codes));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _fallbackExecdClient = fallbackExecdClient;
        _logger.LogDebug("Code interpreter initialized for sandbox: {SandboxId}", sandbox.Id);
    }

    /// <summary>
    /// Creates a new code interpreter from an existing sandbox.
    /// </summary>
    /// <remarks>
    /// By default a strict health check runs before the interpreter is returned: the code
    /// execution service (execd) must answer <c>GET /ping</c> AND the code interpreter
    /// runtime (Jupyter kernel gateway) must be serving, both within
    /// <see cref="CodeInterpreterCreateOptions.ReadyTimeoutSeconds"/>. execd starts
    /// serving before the runtime launches, so the daemon ping alone is not enough.
    /// Set <see cref="CodeInterpreterCreateOptions.SkipHealthCheck"/> to opt out.
    /// </remarks>
    /// <param name="sandbox">The sandbox to wrap.</param>
    /// <param name="options">Optional creation options.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>A new code interpreter instance.</returns>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="sandbox"/> is null.</exception>
    /// <exception cref="SandboxException">Thrown when endpoint discovery or adapter initialization fails.</exception>
    /// <exception cref="SandboxReadyTimeoutException">Thrown when the code execution service health check times out.</exception>
    public static async Task<CodeInterpreter> CreateAsync(
        Sandbox sandbox,
        CodeInterpreterCreateOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        if (sandbox == null)
        {
            throw new InvalidArgumentException("sandbox cannot be null");
        }

        var loggerFactory = options?.Diagnostics?.LoggerFactory ?? sandbox.SharedLoggerFactory ?? NullLoggerFactory.Instance;
        var logger = loggerFactory.CreateLogger("OpenSandbox.CodeInterpreter.CodeInterpreter");
        var endpoint = await sandbox.GetEndpointAsync(Constants.DefaultExecdPort, cancellationToken).ConfigureAwait(false);
        logger.LogInformation("Creating code interpreter for sandbox: {SandboxId}", sandbox.Id);
        var protocol = sandbox.ConnectionConfig.Protocol == ConnectionProtocol.Https ? "https" : "http";
        var execdBaseUrl = $"{protocol}://{endpoint.EndpointAddress}";
        var execdHeaders = MergeHeaders(sandbox.ConnectionConfig.Headers, endpoint.Headers);
        var adapterFactory = options?.AdapterFactory ?? DefaultCodeInterpreterAdapterFactory.Create();

        var codes = adapterFactory.CreateCodes(new CreateCodesStackOptions
        {
            ConnectionConfig = sandbox.ConnectionConfig,
            ExecdBaseUrl = execdBaseUrl,
            ExecdHeaders = execdHeaders,
            HttpClientProvider = sandbox.SharedHttpClientProvider,
            LoggerFactory = loggerFactory
        });

        // Fallback execd probe for custom codes adapters that do not implement
        // IExecdHealth; shares the sandbox's HTTP client.
        var fallbackExecdClient = new HttpClientWrapper(
            sandbox.SharedHttpClientProvider.HttpClient,
            execdBaseUrl,
            execdHeaders,
            loggerFactory.CreateLogger("OpenSandbox.CodeInterpreter.HttpClientWrapper"));

        var interpreter = new CodeInterpreter(sandbox, codes, logger, fallbackExecdClient);

        if (!(options?.SkipHealthCheck ?? false))
        {
            await interpreter.WaitUntilReadyAsync(new CodeInterpreterWaitUntilReadyOptions
            {
                ReadyTimeoutSeconds = options?.ReadyTimeoutSeconds ?? Constants.DefaultReadyTimeoutSeconds,
                PollingIntervalMillis = options?.HealthCheckPollingInterval ?? Constants.DefaultHealthCheckPollingIntervalMillis,
            }, cancellationToken).ConfigureAwait(false);
        }

        return interpreter;
    }

    /// <summary>
    /// Checks whether the code interpreter is healthy (strict check).
    /// </summary>
    /// <remarks>
    /// Healthy means both:
    /// <list type="bullet">
    /// <item>the code execution service (execd) answers <c>GET /ping</c>; and</item>
    /// <item>the code interpreter runtime (Jupyter kernel gateway) is serving
    /// inside the sandbox, verified by probing its listen port through the
    /// execd command API.</item>
    /// </list>
    /// Exceptions from either leg are treated as unhealthy.
    /// </remarks>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>True if the code interpreter is healthy, false otherwise.</returns>
    public async Task<bool> IsHealthyAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            if (!await PingExecdAsync(cancellationToken).ConfigureAwait(false))
            {
                return false;
            }

            return await IsRuntimeServingAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.LogDebug(ex, "Health check failed for code interpreter {SandboxId}", Id);
            return false;
        }
    }

    /// <summary>
    /// Pings the execd daemon on the interpreter's own endpoint. Prefers the codes
    /// service's optional <see cref="IExecdHealth"/> capability; custom adapters that
    /// do not implement it fall back to a direct probe sharing the sandbox HTTP client.
    /// </summary>
    private async Task<bool> PingExecdAsync(CancellationToken cancellationToken)
    {
        if (Codes is IExecdHealth health)
        {
            return await health.PingAsync(cancellationToken).ConfigureAwait(false);
        }

        if (_fallbackExecdClient == null)
        {
            return false;
        }

        try
        {
            await _fallbackExecdClient.GetAsync("/ping", cancellationToken: cancellationToken).ConfigureAwait(false);
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogDebug(ex, "Fallback execd ping failed for code interpreter {SandboxId}", Id);
            return false;
        }
    }

    private async Task<bool> IsRuntimeServingAsync(CancellationToken cancellationToken)
    {
        try
        {
            var execution = await Sandbox.Commands.RunAsync(
                CodeInterpreterHealthCheck.RuntimeCheckCommand,
                cancellationToken: cancellationToken).ConfigureAwait(false);
            return execution?.Error == null;
        }
        catch (Exception ex)
        {
            _logger.LogDebug(ex, "Runtime check failed for code interpreter {SandboxId}", Id);
            return false;
        }
    }

    /// <summary>
    /// Waits until the strict health check passes (execd answers <c>GET /ping</c>
    /// and the interpreter runtime process is alive).
    /// </summary>
    /// <param name="options">The wait options.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <exception cref="SandboxReadyTimeoutException">Thrown when the health check times out.</exception>
    /// <exception cref="OperationCanceledException">Thrown when <paramref name="cancellationToken"/> is canceled.</exception>
    public async Task WaitUntilReadyAsync(
        CodeInterpreterWaitUntilReadyOptions options,
        CancellationToken cancellationToken = default)
    {
        _logger.LogDebug(
            "Start readiness check for code interpreter {SandboxId} (timeoutSeconds={TimeoutSeconds})",
            Id, options.ReadyTimeoutSeconds);
        var timeout = TimeSpan.FromSeconds(options.ReadyTimeoutSeconds);
        var stopwatch = Stopwatch.StartNew();
        var attempt = 0;
        var errorDetail = "Health check returned false continuously.";

        while (true)
        {
            cancellationToken.ThrowIfCancellationRequested();

            if (stopwatch.Elapsed > timeout)
            {
                throw new SandboxReadyTimeoutException(
                    $"Code interpreter {Id} health check timed out after {options.ReadyTimeoutSeconds}s ({attempt} attempts). " +
                    $"{errorDetail} The code execution service (execd) or the interpreter runtime (Jupyter) " +
                    "did not become ready. Set SkipHealthCheck to skip this check.");
            }
            attempt++;

            try
            {
                if (await IsHealthyAsync(cancellationToken).ConfigureAwait(false))
                {
                    _logger.LogInformation("Code interpreter is ready: {SandboxId}", Id);
                    return;
                }

                errorDetail = "Health check returned false continuously.";
            }
            catch (Exception ex)
            {
                _logger.LogDebug(ex, "Readiness probe failed for code interpreter {SandboxId}", Id);
                errorDetail = $"Last health check error: {ex.Message}";
            }

            var remaining = timeout - stopwatch.Elapsed;
            if (remaining <= TimeSpan.Zero)
            {
                continue;
            }

            var pollingInterval = TimeSpan.FromMilliseconds(options.PollingIntervalMillis);
            var delay = pollingInterval < remaining ? pollingInterval : remaining;
            await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
        }
    }

    private static IReadOnlyDictionary<string, string> MergeHeaders(
        IReadOnlyDictionary<string, string> baseHeaders,
        IReadOnlyDictionary<string, string>? overrideHeaders)
    {
        var merged = baseHeaders.ToDictionary(header => header.Key, header => header.Value);
        if (overrideHeaders != null)
        {
            foreach (var header in overrideHeaders)
            {
                merged[header.Key] = header.Value;
            }
        }

        return merged;
    }
}
